crates/task-impls/src/transactions.rs

```diff
@@ -0,0 +1,346 @@
+            SequencingHotShotEvent::Shutdown => {
+                return Some(HotShotTaskCompleted::ShutDown);
+            }
+            _ => {}
+        }
+        None
+    }
+
+    #[instrument(skip_all, fields(id = self.id, view = *self.cur_view), name = "Transaction Handling Task", level = "error")]
+    async fn wait_for_transactions(
+        &self,
+        parent_leaf: SequencingLeaf<TYPES>,
+    ) -> Option<Vec<TYPES::Transaction>> {
+        let task_start_time = Instant::now();
+
+        // let parent_leaf = self.parent_leaf().await?;
+        let previous_used_txns = match parent_leaf.deltas {
+            Either::Left(block) => block.contained_transactions(),
+            Either::Right(_commitment) => HashSet::new(),
+        };
+
+        let consensus = self.consensus.read().await;
+
+        let receiver = consensus.transactions.subscribe().await;
+
+        loop {
+            let all_txns = consensus.transactions.cloned().await;
+            debug!("Size of transactions: {}", all_txns.len());
+            let unclaimed_txns: Vec<_> = all_txns
+                .iter()
+                .filter(|(txn_hash, _txn)| !previous_used_txns.contains(txn_hash))
+                .collect();
+
+            let time_past = task_start_time.elapsed();
+            if unclaimed_txns.len() < self.api.min_transactions()
+                && (time_past < self.api.propose_max_round_time())
+            {
+                let duration = self.api.propose_max_round_time() - time_past;
+                let result = async_timeout(duration, receiver.recv()).await;
+                match result {
+                    Err(_) => {
+                        // Fall through below to updating new block
+                        error!(
+                            "propose_max_round_time passed, sending transactions we have so far"
+                        );
+                    }
+                    Ok(Err(e)) => {
+                        // Something unprecedented is wrong, and `transactions` has been dropped
+                        error!("Channel receiver error for SubscribableRwLock {:?}", e);
+                        return None;
+                    }
+                    Ok(Ok(_)) => continue,
+                }
+            }
+            break;
+        }
+        let all_txns = consensus.transactions.cloned().await;
+        let txns: Vec<TYPES::Transaction> = all_txns
+            .iter()
+            .filter_map(|(txn_hash, txn)| {
+                if previous_used_txns.contains(txn_hash) {
+                    None
+                } else {
+                    Some(txn.clone())
+                }
+            })
+            .collect();
+        Some(txns)
+    }
+
+    /// Event filter for the transaction task
+    pub fn filter(event: &SequencingHotShotEvent<TYPES, I>) -> bool {
+        matches!(
+            event,
+            SequencingHotShotEvent::TransactionsRecv(_)
+                | SequencingHotShotEvent::DAVoteRecv(_)
```

Comment:
I think `DAVoteRecv` should be replaced by `LeafDecided`.

Reply:
Yes good catch, took me a second to realize why the test were passing.  We only remove transactions on decide so we might get duplicates with this bug left in but still will be able to form blocks with transactions.  Fixed now